### PR TITLE
Drop support for Python 2.7 and 3.5.

### PR DESCRIPTION
**Pull request by icemac**

### CODE REVIEW

**Comment by dataflake** [comment]
`.travis.yml` and `setup.py` still refer to Python 2

File: setup.py [hunk 1]
```diff
@@ -36,13 +36,15 @@
         "Programming Language :: Python :: 2",
         "Programming Language :: Python :: 2.7",
```
Comment (dataflake) [inline setup.py]: Can go away

File: setup.py [hunk 2]
```diff
@@ -36,13 +36,15 @@
...
-        "Programming Language :: Python :: 3.5",
         "Programming Language :: Python :: 3.6",
         "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
         "Programming Language :: Python :: Implementation :: CPython",
         "Programming Language :: Python :: Implementation :: PyPy",
         "Topic :: System :: Systems Administration :: Authentication/Directory :: LDAP",  # noqa: E501
     ],
+    python_requires='>=2.7,!=3.0.*,!=3.1.*,!=3.2.*,!=3.3.*,!=3.4.*,!=3.5.*',
```
Comment (dataflake) [inline setup.py]: Can change to `'>=3.6'`

File: setup.py [hunk 3]
```diff
@@ -36,13 +36,15 @@
...
     install_requires=[
         'six',
     ],
```
Comment (dataflake) [inline setup.py]: Can probably go

File: .travis.yml [hunk 4]
```diff
@@ -1,17 +1,11 @@
 # custom
 language: python
-sudo: false
 python:
-    - 2.7
```
Comment (mgedmin) [inline .travis.yml:3]: If you're dropping 2.7, please mention that in the PR title.

(Also, I thought there were a few people who wanted to keep maintaining 2.7 support for ZF packages, and the consensus was we weren't going to make their life harder without a good reason?)

File: src/AuthEncoding/AuthEncoding.py [hunk 5]
```diff
@@ -39,7 +37,7 @@ def _reseed():
         # properties of the chosen random sequence slightly, but this
         # is better than absolute predictability.
         random.seed(sha256(
-            "%s%s%s" % (random.getstate(), time.time(), getpid())
+            "{}{}{}".format(random.getstate(), time.time(), getpid())
```
Comment (mgedmin) [inline src/AuthEncoding/AuthEncoding.py]: Why not an f-string if we're going hard for >= 3.6?

File: src/AuthEncoding/AuthEncoding.py [hunk 6]
```diff
@@ -62,7 +74,7 @@ def constant_time_compare(val1, val2):
     if len(val1) != len(val2):
         return False
     result = 0
-    for x, y in zip(six.iterbytes(val1), six.iterbytes(val2)):
+    for x, y in zip(iter(val1), iter(val2)):
```
Comment (mgedmin) [inline src/AuthEncoding/AuthEncoding.py]: I 
```suggestion
    for x, y in zip(val1, val2):
```

File: src/AuthEncoding/AuthEncoding.py [hunk 7]
```diff
@@ -110,7 +122,7 @@ def generate_salt(self):
         # All 256 characters are available.
         salt = b''
         for n in range(7):
-            salt += six.int2byte(_randrange(256))
+            salt += bytes((_randrange(256),))
         return salt
```
Comment (mgedmin) [inline src/AuthEncoding/AuthEncoding.py:126]: I'd replace this whole method with `return os.urandom(7)` (in a separate PR).

**Review by mgedmin — APPROVED** [review]
lgtm, assuming we're fine with dropping 2.7 as a group

(I wanted to link to a previous discussion about that but couldn't find it)

**Comment by dataflake** [comment]
> lgtm, assuming we're fine with dropping 2.7 as a group

Michael worked on several packages yesterday and wanted to just drop 3.5 first, because Zope 5 doesn't support it. But he overlooked that dropping 3.5 would make a package potentially incompatible with Zope 4.x, which explicitly supports it. Then we both reasoned that he could just make the full step of dropping 2.7 as well, so the package would end up for Zope 5 only. But Jason Madden spoke up about some packages where he definitely wants to retain 2.7.

I'm now thinking that we should stop such efforts to drop compatibility unless there are really good reasons and revisit it when Zope 5 has been around a bit and adopted by the bigger "customers" in the ecosystem, e.g. Plone.

**Comment by dataflake** [comment]
To get this out of the critical path for a Zope 5 release I have pulled out the changes that add support for Python 3.8 and Python 3.9 and merged those into `master`. I then published release 4.2, which we can use for Zope 5 (and 4).

This PR only contains the changes for dropping Python 2 and Python 3.5 support, and that should wait until we have consensus on when to drop this support for packages where keeping that support doesn't introduce a big burden. I added the `do not merge` tag for that reason.

**Comment by dataflake** [comment]
Superceded by #14

**Pull request closed without merging**